src/Microsoft.Restier.WebApi/Query/ODataCountOption.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Restier.WebApi.Query
+{
+    internal class ODataCountOption
+    {
+        /// <summary>
+        /// Gets or sets a value indicating whether the total
+        /// number of items should be retrieved when the
+        /// result has been filtered using paging operators.
+        /// </summary>
+        /// <remarks>
+        /// Setting this to <c>true</c> may have a performance impact as
+        /// the data provider may need to execute two independent queries.
+        /// </remarks>
+        public bool IncludeTotalCount { get; set; }
+
+        public Action<long> SetTotalCount { get; set; }
+    }
+}
```

Comment:
From Restier view, I see request to expand both provider part (Like mongoDB) and publisher part (Like Swagger),, but I do expect we will have more request on the provider part to support more data source to be exposed as OData service, and much less on the publisher part.
Also I was thinking different publisher may have different logic, like for count, OData does not consider top/skip but consider filter/search, for other protocol, it may not consider top, but consider skip/filter/search.
 
Based on this assumption, I choose option 2.  And if we do not want to introduce more logic into ApiContext, then may we have  class to be ODataQueryOption or similar. 
You are expert on this code (I just know the code about 2 weeks :)), you can make the call.
I will discuss with Lewis when he is back.